### PR TITLE
ci upgrade

### PR DESCRIPTION
**Pull request by mpetuska**
- Multiplatform tests
- Automatic snapshot version releases and bumps
- Full releases are now triggered by creating a GH release with tag `x.y.z`
- Dokka publication to gh pages on full release

### CODE REVIEW

**Comment by mpetuska** [comment]
Before merging this will need branch protection rule update to expect the new builds.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    strategy:
+      matrix:
+        repository:
+          - name: Github Packages
+            tasks: publishAllPublicationsToGitHubRepository
+            enabled: ${{ github.event.inputs.skipGitHub != 'y' }}
+          - name: Maven Central
+            # `closeSonatypeStagingRepository` Requires to manually release a new version on https://s01.oss.sonatype.org/#stagingRepositories
+            # If you're brave, you could change it to `closeAndReleaseSonatypeStagingRepository` to fully automate the release
+            tasks: publishToSonatype closeAndReleaseSonatypeStagingRepository
```
Comment (patjackson52) [inline .github/workflows/release.yml:105]: If this can fully automate release publishing I'm ok with that for now.  Potentially risk publishing busted artifacts, but given we are all busy I think it's worth the trade off for now.

File: README.md [hunk 2]
```diff
@@ -144,11 +167,17 @@ Trello boards - https://trello.com/reduxkotlinorg
...
+
 [badge-jvm]: http://img.shields.io/badge/platform-jvm-orange.svg?style=flat
-[badge-linux]: http://img.shields.io/badge/platform-linux-important.svg?style=flat 
+
+[badge-linux]: http://img.shields.io/badge/platform-linux-important.svg?style=flat
+
 [badge-windows]: http://img.shields.io/badge/platform-windows-informational.svg?style=flat
+
 [badge-mac]: http://img.shields.io/badge/platform-macos-lightgrey.svg?style=flat
-[badge-wasm]: https://img.shields.io/badge/platform-wasm-darkblue.svg?style=flat
```
Comment (patjackson52) [inline README.md:154]: is wasm no longer a target?

**Review by patjackson52 — COMMENTED** [review]
tl;dr of automatic version bumps? snapshots are published to github packages & releases are github packages & maven central?
Would be nice to capture in docs.  Nice addition of gh pages.

File: README.md (hunk 2 above)
Comment (mpetuska) [reply README.md:154]: No since wasm32 was deprecated in kotlin 1.8.0 and new wasm backend is not announced yet

File: .github/workflows/release.yml (hunk 1 above)
Comment (mpetuska) [reply .github/workflows/release.yml:105]: I've added `publishToLoacal` task to our check runs to at least be sure we can build all the artefacts fine. Let's keep it until the first error and discuss alternatives if that happens.

**Comment by mpetuska** [comment]
> tl;dr of automatic version bumps?

In short it basically works as follows:
- When stuff is committed/merged to master or a new tag is created (via gh release) `release` build runs
- Within `release` build there's a call to `resolve-version` build
- `resolve-version` build reads the ref, tag and gradle.properties and figures out the desired version
- If the desired version is SNAPSHOT (i.e. gradle.properties version won) it publishes snapshot builds to gh packages and maven central
- If the version is NOT snapshot (i.e. tag ref won)
  - It publishes release builds to GH packages and maven central.
  - It bumps patch version of the resolved version and makes it SNAPSHOT (i.e. x.y.<z+1>-SNAPSHOT)
  - Then it writes that new SNAPSHOT version to gradle.properties and pushes it back to the master branch

There are some more minor caveats to this, but that's how it works in a nutshell.

**Review by patjackson52 — APPROVED** [review]
💯

**Comment by patjackson52** [comment]
> > Before merging this will need branch protection rule update to expect the new builds.
> 
> @patjackson52 can you sort this out, please? Old test and ktlint builds will never start on this PR because they're no longer there 😁

updated rules - manually merged this one as the checks weren't updating - everything passed
